src/features/courses/course-controller.ts

```diff
@@ -3268,6 +3270,54 @@ class CourseController {
         });
     };
 
+    async createAttachment({
+        obj,
+        studentGradeId,
+        studentGradeInstanceId
+    }: CreateAttachmentOptions): Promise<ProblemAttachment> {
+        return useDatabaseTransaction(async (): Promise<ProblemAttachment> => {
+            if (_.isNil(studentGradeId) === _.isNil(studentGradeInstanceId)) {
+                throw new IllegalArgumentException('Student grade XOR student grade instance id must be supplied');
+            }
+
+            if(!_.isNil(studentGradeInstanceId)) {
+                if(!_.isNil(studentGradeId)) {
+                    throw new IllegalArgumentException('This should not be possible since _.isNil student grade id and student grade instance id cannot be equal');
+                }
+                const studentGradeInstance = await courseRepository.getStudentGradeInstance({
+                    id: studentGradeInstanceId
+                });
+                studentGradeId = studentGradeInstance.studentGradeId;
+            }
+
+            const result = await courseRepository.createAttachment(obj);
+
+            /**
+             * Currently this could and should be validated by the constraint
+             * however that causes gaps in ids which I know we aren't too worried about
+             * furthermore we will need to fetch the grade anyway for permissions so why not throw it in for now
+             */
+            if(_.isNil(studentGradeId)) throw new RederlyError('Based on above logic this is not possible, however use strict did not pick up on that');
+            // on failure this will throw an error and bubble up
+            await courseRepository.getStudentGrade({
```

Comment:
What failure is expected here? Also, if this is just a nil check, should it happen in `createStudentGradeProblemAttachment`?